File: src/main/java/net/yakclient/mixin/api/Injection.java
package net.yakclient.mixin.api;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * {@code @Injection} marks a method as a Injection, who's Java
 * bytecode will be injected into the reciprocal.
 *
 * This annotation is meant to be very simple and easy to use.
 * In fact nothing has to be pre-defined and values will
 * default to acceptable options.
 *
 * A basic example, in this case all we have to provide is
 * the annotation;
 *
 * <pre>
 * {@code
 *  ...
 * @Injection
 *     public void myMethod(Integer... params) { ... }
 *  ...
 * }
 * </pre>
 *
 * In a more complex example we could provide as many
 * options as we would like;
 *
 * <pre>
 * {@code
 *  ...
 *     @Injection(to = "a_different_name", type = InjectionType.BEFORE_INVOKE, priority = Priority.LOW)
 *     public void myMethod(Integer... params) { ... }
 *  ...
 * }
 * </pre>
 *
 * @see InjectionType
 * @see Priority
 *
 * @author Durgan McBroom
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
public @interface Injection {
    String METHOD_SELF = "<SELF>";

    /**
     * The {@code Injection#to()} method defines the method that
     * will be injected into. If this value is not provided it will
     * default to the name of the method that called.
     *
     * @return the method that will be mixed into.
     */
    String to() default METHOD_SELF;

    /**
     * The {@code Injection#type()} method specifies the type
     * of injection that will be performed. This will default
     * to {@code InjectionType.AFTER_BEGIN}
     *
     * @see InjectionType
     *
     * @return the type of injection that will be performed.
     */
    InjectionType type() default InjectionType.AFTER_BEGIN;

    /**
     * Specifies a injection priority that will in the end determine
     * what different injections will go first if there is a
     * collision.
     *
     * @return the priority of this injection.
     */
    int priority() default Priority.MEDIUM;
}

File: api/src/main/kotlin/net/yakclient/mixins/api/Injection.kt
package net.yakclient.mixins.api

/**
 * `@Injection` marks a method as an Injection. In a mixin this will be
 * code that is injected into another spot.
 *
 * This annotation is meant to be very simple and easy to use.
 * In fact nothing has to be pre-defined and values will
 * default to acceptable options.
 *
 * The most basic example is as follows:
 * ```
 * @Injection
 * fun doSomething(int: Int, any: Any) : String { /*Injection*/ }
 * ```
 * This configuration will inject the body of the annotated method into
 * the methods with the signature `doSomething(ILkotlin/Any;)Ljava/lang/String`
 *
 * You can of course change options to something more like this:
 *
 * ```
 * @Injection(
 *    to="doSomething(ILkotlin/Any;)",
 *    type = InjectionType.AFTER_BEGIN,
 *    priority = Priority.MEDIUM)
 * fun asdf(int: Int, any: Any) /* Can have a void return type where the mixin does not exit the method */ { /*Injection*/ }
 * ```
 *
 * This will inject to a method with the signature `doSomething(ILkotlin/Any;)`(ignoring return type) with a
 * injection type of [AFTER_BEGIN][InjectionType.AFTER_BEGIN] and a [medium priority][Priority.MEDIUM]
 *
 * @see InjectionType
 * @see Priority
 *
 * @author Durgan McBroom
 */

public const val METHOD_SELF: String = "<SELF>"

@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
public annotation class Injection(
    /**
     * The `Injection#to()` method defines the method that
     * will be injected into. If this value is not provided it will
     * default to the name of the method that called.
     *
     * @return the method that will be mixed into.
     */
    val to: String = METHOD_SELF,
    /**
     * The `Injection#type()` method specifies the type
     * of injection that will be performed. This will default
     * to `InjectionType.AFTER_BEGIN`
     *
     * @see InjectionType
     *
     *
     * @return the type of injection that will be performed.
     */
    val type: Int = InjectionType.AFTER_BEGIN,
    /**
     * Specifies a injection priority that will in the end determine
     * what different injections will go first if there is a
     * collision.
     *
     * @see Priority
     *
     * @return the priority of this injection.
     */
    val priority: Int = Priority.MEDIUM
)